nmdc_schema/migrators/migration_reporter.py

```diff
@@ -0,0 +1,261 @@
+                              key=lambda x: (x[0][0], x[0][1], -(x[1]['conformant'] + x[1]['non_conformant'])))
+        
+        for (class_name, slot_name, subclass_type, source_value, target_value), counts in sorted_updates:
+            conformant_count = counts['conformant']
+            non_conformant_count = counts['non_conformant']
+            updated_count = counts['updated']
+            
+            # If subclass type is the same as class, leave it blank
+            display_subclass = "" if subclass_type == class_name else subclass_type
+            
+            row = (f"{class_name:<30} {display_subclass:<25} {slot_name:<25} "
+                   f"{source_value:<15} {conformant_count:<12} {non_conformant_count:<12} {updated_count:<8} {target_value:<15}")
+            self.logger.info(row)
+        
+        self.logger.info("")
+    
+    def _generate_unmapped_values_table(self) -> None:
+        """Generate the unmapped values table."""
+        self.logger.info("UNMAPPED VALUES")
+        self.logger.info("-" * 40)
+        
+        if not self.unmapped_values:
+            self.logger.info("No unmapped values found.")
+            self.logger.info("")
+            return
+        
+        # Table header
+        self.logger.info(f"{'Class.Slot':<35} {'Unmapped Value':<20} {'Total Count':<12}")
+        self.logger.info("-" * 67)
+        
+        # Sort by class.slot, then by count (descending)
+        for class_slot in sorted(self.unmapped_values.keys()):
+            values = self.unmapped_values[class_slot]
+            sorted_values = sorted(values.items(), key=lambda x: -x[1])
+            
+            for value, count in sorted_values:
+                self.logger.info(f"{class_slot:<35} {value:<20} {count:<12}")
+        
+        self.logger.info("")
+    
+    def _generate_missing_values_table(self) -> None:
+        """Generate the missing values table."""
+        self.logger.info("MISSING VALUES")
+        self.logger.info("-" * 40)
+        
+        if not self.missing_values:
+            self.logger.info("No missing values found.")
+            self.logger.info("")
+            return
+        
+        # Table header
+        self.logger.info(f"{'Class.Slot':<35} {'Total Missing':<12}")
+        self.logger.info("-" * 47)
+        
+        # Sort by class.slot
+        for class_slot in sorted(self.missing_values.keys()):
+            count = self.missing_values[class_slot]
+            self.logger.info(f"{class_slot:<35} {count:<12}")
+        
+        self.logger.info("")
+    
+
+
+# Generic schema path utilities for migration reporting
+def parse_schema_path(path: str) -> List[str]:
+    """
+    Parses a document path into schema-relevant components, filtering out array indices.
+
+    Args:
+        path: Path like "substances_used[0].volume" or "extraction.input_mass"
+
+    Returns:
+        List of schema slot names: ["substances_used", "volume"] or ["extraction", "input_mass"]
+    """
```

Comment:
I suggest adding some doctests to this helper function (I confirmed these pass).

```suggestion

    >>> parse_schema_path("extraction.input_mass")  # no array
    ['extraction', 'input_mass']
    >>> parse_schema_path("substances_used[0].volume")  # 1-D array
    ['substances_used', 'volume']
    """
```